src/override/rust.cc

```diff
@@ -9,47 +9,183 @@
 
 using namespace snmalloc;
 
-extern "C" SNMALLOC_EXPORT void*
-  SNMALLOC_NAME_MANGLE(rust_alloc)(size_t alignment, size_t size)
+namespace
 {
-  return ThreadAlloc::get().alloc(aligned_size(alignment, size));
+  enum class SizeChange
+  {
+    Unknown,
+    Grow,
+    Shrink
+  };
+  template<ZeroMem zero_mem, SizeChange size_change>
+  SNMALLOC_FAST_PATH_INLINE void* realloc_helper(
+    Alloc& alloc,
+    void* ptr,
+    size_t old_alignment,
+    size_t old_size,
+    size_t new_alignment,
+    size_t new_size)
+  {
+    size_t aligned_old_size = aligned_size(old_alignment, old_size),
+           aligned_new_size = aligned_size(new_alignment, new_size);
+    if (
+      size_to_sizeclass_full(aligned_old_size).raw() ==
+      size_to_sizeclass_full(aligned_new_size).raw())
+      return ptr;
+    // this may memset more than it needs, but doing so increases
+    // the possibility for large area to be efficiently zeroed by special
+    // methods
+    void* p = alloc.alloc<zero_mem>(aligned_new_size);
+    if (p)
+    {
+      if constexpr (size_change == SizeChange::Unknown)
+        std::memcpy(p, ptr, old_size < new_size ? old_size : new_size);
+      else if constexpr (size_change == SizeChange::Grow)
+        std::memcpy(p, ptr, old_size);
```

Comment:
Should we have an assert here, so that in a Debug build we can detect incorrect usage?